### ISSUE TITLE
libispc - ISPC in library form

### ISSUE THREAD

**Issue by breakin** [ispc/ispc#791]
A feature request! (Sorry if you've heard it before... I know I requested it from Matt Pharr when ISPC was new...)

In addition to the ISPC-compiler a C-library could be created. It would use the JIT capabilities of LLVM to let the library user synthesize and execute programs. This would be useful when using ISPC as a shader system in an application. This setup would not work for everyone since there are security implications, but it would be super useful for some.

The library would need to:
- Receive ISPC-sources (preferably no disk access required)
- Enumerate exported ISPC-function (and their arguments)
- Let the user get function pointer to an exported (and JIT:ed) ISPC-function so he can call it
- Be a standalone C-API, pre-compiled on all major platforms with no dependencies (on say dynamic CRT).

The hard part as I see it is in enumerating the argument to the functions. It would need a way to "talk about" all the possible arguments to exported ISPC-functions.

I realize that using ispc.exe with emit-llvm and then hooking it up with LLVM yourself it possible, but that is hard work for most end-users (especially if they are multi-platform).

**Comment by inducer** [comment]
:+1:

**Comment by hlide** [comment]
+1. This is the big reason I do not use ISPC.

**Comment by capr** [comment]
ISPC in lib form would be cool. Building it to dynamically link against an existing LLVM shared lib would be even cooler. This would allow me to use both [terra](http://terralang.org/) and ispc in the same program without duplicating (not to mention recompiling) LLVM. Thanks for thinking about this.

**Comment by dbabokin** [comment]
Dynamically linking against existing LLVM libs is not possible on practice, as ISPC is tightly coupled with particular version of LLVM libs (LLVM API is not stable). We have ifdefs to be able to build with different LLVM versions. Also we sometimes apply patches to LLVM that we use to make it work well for us.

**Comment by capr** [comment]
Yes, and yet it's the only way to have multiple pieces of software that depend on LLVM in the same app (as long as you build them all against the exact same version of LLVM, I understand that). As for patches, are those API-breaking? Is there anything that can be done about it? Adding LLVM twice to an app is a 25 MB hit to the binary.

**Comment by dbabokin** [comment]
Patches are only for stability. We are trying to avoid them, but we have to back-port the fixes to already release versions of LLVM.

Anyway, deciding on static vs dynamic linking is lower order problem than building ispc as library.

**Comment by damiendr** [comment]
I would also very much like this. I use ISPC as a JIT target and the CLI overhead is an issue.

**Comment by nlguillemot** [comment]
Maybe it's a separate issue, but this might also be helpful for implementing intellisense for ISPC.
